lib/spack/spack/solver/asp.py

```diff
@@ -3065,6 +3065,19 @@ def setup(
         self.possible_compilers = list(candidate_compilers)
         self.possible_compilers.sort()  # type: ignore[call-overload]
 
+        should_mix = spack.config.get("concretizer:compiler_mixing", True)
+        if isinstance(should_mix, bool) and should_mix:
+            pass
+        else:
+            for lang in ["c", "cxx", "fortran"]:
+                self.gen.fact(fn.no_compiler_mixing(lang))
+
+            try:
+                for pkg_name in should_mix:
+                    self.gen.fact(fn.allow_mixing(pkg_name))
+            except TypeError:
+                pass
```

Comment:
```suggestion
        if should_mix is not True:  # anything besides should_mix: true
            for lang in ["c", "cxx", "fortran"]:
                self.gen.fact(fn.no_compiler_mixing(lang))

        if isinstance(should_mix, list):  # user specified an allow-list
            for pkg_name in should_mix:
                self.gen.fact(fn.allow_mixing(pkg_name))
```
I don't think there's anything logically wrong with the current implementation, but it's a bit hard to read. I'm pretty sure this is logically equivalent.